polynote-server/src/main/scala/polynote/server/KernelPublisher.scala

```diff
@@ -158,10 +162,31 @@ class KernelPublisher private (
     } yield ()
   }.ignore
 
-  def subscribe(): RIO[BaseEnv with GlobalEnv with PublishMessage, KernelSubscriber] = for {
-    subscriberId       <- ZIO.effectTotal(nextSubscriberId.getAndIncrement())
-    subscriber         <- KernelSubscriber(subscriberId, this)
-  } yield subscriber
+  def subscribe(): RIO[BaseEnv with GlobalEnv with PublishMessage, KernelSubscriber] = subscribing.withPermit {
+    for {
+      isClosed           <- closed.isDone
+      _                  <- if (isClosed) ZIO.fail(PublisherClosed) else ZIO.unit
```

Comment:
hm, are there circumstances under which we might try to subscribe on an already-closed publisher?

Reply:
Well, if user `A` closes the last subscriber just as user `B` starts subscribing, I could see that being a race condition. A very unlikely one, but it doesn't hurt to check.